zebra-state/src/service/check/nullifier.rs

```diff
@@ -56,40 +56,33 @@ pub(crate) fn no_duplicates_in_finalized_chain(
     Ok(())
 }
 
-/// Reject double-spends of nullifiers:
-/// - one from this [`Transaction`], and the other already committed to the
-///   provided [`ZebraDb`].
-///
-/// # Consensus
-///
-/// > A nullifier MUST NOT repeat either within a transaction,
-/// > or across transactions in a valid blockchain.
-/// > Sprout and Sapling and Orchard nullifiers are considered disjoint,
-/// > even if they have the same bit pattern.
-///
-/// <https://zips.z.cash/protocol/protocol.pdf#nullifierset>
-// TODO: impl nullifier getters for a trait and use a generic type to unify
-//       this fn with the one for `PreparedBlock`
-#[tracing::instrument(skip_all)]
-fn tx_no_duplicates_in_finalized_chain(
-    finalized_chain: &ZebraDb,
-    transaction: &Arc<Transaction>,
-) -> Result<(), ValidateContextError> {
-    for nullifier in transaction.sprout_nullifiers() {
-        if finalized_chain.contains_sprout_nullifier(nullifier) {
-            Err(nullifier.duplicate_nullifier_error(true))?;
-        }
-    }
-
-    for nullifier in transaction.sapling_nullifiers() {
-        if finalized_chain.contains_sapling_nullifier(nullifier) {
-            Err(nullifier.duplicate_nullifier_error(true))?;
+/// Returns a Ok(()) if the predicate(s) return false for every revealed nullifier
+/// Returns Err(DuplicateNullifierError) for the first nullifier that matches the predicate(s)
```

Comment:
Nitpick: The non-finalized state contains multiple chain forks, so we should probably say "chain" rather than "state" throughout this function.

Can you also document each of the arguments here?

```suggestion
/// Returns `Err(DuplicateNullifierError)` if any of the `revealed_nullifiers` are found in the non-finalized or finalized chains.
/// Returns `Ok(())` if all the `revealed_nullifiers` have not been seen in either chain.
```